core/src/View/Kokkos_BasicView.hpp

```diff
@@ -0,0 +1,611 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+static_assert(false,
+              "Including non-public Kokkos header files is not allowed.");
+#endif
+
+#ifndef KOKKOS_BASIC_VIEW_HPP
+#define KOKKOS_BASIC_VIEW_HPP
+#include <type_traits>
+#include <Kokkos_Macros.hpp>
+#include <impl/Kokkos_Utilities.hpp>
+#include <impl/Kokkos_SharedAlloc.hpp>
+#include "Kokkos_ViewAlloc.hpp"
+#include "MDSpan/Kokkos_MDSpan_Header.hpp"
+#include "MDSpan/Kokkos_MDSpan_Accessor.hpp"
+#include "MDSpan/Kokkos_MDSpan_Layout.hpp"
+#include <View/Kokkos_ViewCtor.hpp>
+#include <View/Kokkos_ViewTraits.hpp>
+
+#define KOKKOS_NO_UNIQUE_ADDRESS _MDSPAN_NO_UNIQUE_ADDRESS
+namespace Kokkos {
+
+namespace Impl {
+constexpr inline struct subview_ctor_tag_t {
+} subview_ctor_tag;
+
+template <class T>
+struct KokkosSliceToMDSpanSliceImpl {
+  using type = T;
+  KOKKOS_FUNCTION
+  static constexpr decltype(auto) transform(const T &s) { return s; }
+};
+
+template <>
+struct KokkosSliceToMDSpanSliceImpl<Kokkos::ALL_t> {
+  using type = full_extent_t;
+  KOKKOS_FUNCTION
+  static constexpr decltype(auto) transform(Kokkos::ALL_t) {
+    return full_extent;
+  }
+};
```

Comment:
we probably would need the warning silencer at which point it feels that this thing would become less readable than before ( which I think it kinda already is). Also we may need to transform Kokkos::pair to std::pair if we build against std::mdspan instead of our own implementation in the future.